<launch>

  <arg name="map_yaml"                doc="filepath to previously captured map (required)" />
  <arg name="rviz"     default="true" doc="whether to show Rviz" />

  <!-- STRETCH DRIVER -->
  <param name="/stretch_driver/broadcast_odom_tf" type="bool" value="true"/>
  <param name="/stretch_driver/mode" type="string" value="navigation" />
  <include file="$(find stretch_core)/launch/stretch_driver.launch" pass_all_args="true"/>

  <!-- LASER RANGE FINDER -->
  <!--include file="$(find stretch_core)/launch/rplidar.launch" /-->
  <node name="lidar_node" pkg="rplidar_ros" type="rplidarNode" output="screen">
    <param name="serial_port"         type="string" value="/dev/hello-lrf"/>
    <param name="serial_baudrate"     type="int"    value="115200"/> 
    <param name="frame_id"            type="string" value="laser"/>
    <param name="inverted"            type="bool"   value="false"/>
    <param name="angle_compensate"    type="bool"   value="true"/>
    <param name="scan_mode"           type="string" value="Boost"/>
  </node>
  <!--node pkg="laser_filters" type="scan_to_scan_filter_chain"
      name="laser_filter">
  </node-->

  <!-- hokuyoのLRFを使う場合．stretch_tf.pyの実行も必要． -->
  <!--node name="urg_node" pkg="urg_node" type="urg_node">
    <param name="serial_port" value="/dev/ttyACM6"/>
    <param name="frame_id" value="hokuyo_laer"/>
  </node-->

  <!-- MAP SERVER -->
  <!-- 自己位置推定用とパスプラン用の2つのマップを利用 -->
  <arg name="map_file_loc" default="~/map.yaml"/>
  <node name="map_server_loc" pkg="map_server" type="map_server" args="$(arg map_file_loc)">
  </node>

  <arg name="map_file_pp" default="~/map.yaml"/>
  <node name="map_server_pp" pkg="map_server" type="map_server" args="$(arg map_file_pp)">
  <remap from="map" to="map_pp"/>
  </node>




  <!-- LOCALIZATION -->
  <!--include file="$(find amcl)/examples/amcl_diff.launch" /-->

  <node pkg="amcl" type="amcl" name="amcl">
    <param name="min_particles"             value="500"/>
    <param name="max_particles"             value="3000"/>
    <param name="kld_err"                   value="0.05"/>
    <param name="update_min_d"              value="0.1"/><!-- オドメトリの精度が悪いので頻繁に更新 -->
    <param name="update_min_a"              value="0.1"/>
    <param name="resample_interval"         value="1"/>
    <param name="transform_tolerance"       value="1.0"/>
    <param name="recovery_alpha_slow"       value="0.0"/>
    <param name="recovery_alpha_fast"       value="0.0"/>
    <param name="initial_pose_x"            value="0"/>
    <param name="initial_pose_y"            value="0"/>
    <param name="initial_pose_a"            value="0"/>
    <param name="gui_publish_rate"          value="10.0"/>
    <param name="use_map_topic"             value="false"/>

    <remap from="scan"                      to="scan"/>
    <param name="laser_max_range"           value="12.0"/>
    <param name="laser_max_beams"           value="60"/>
    <param name="laser_z_hit"               value="0.95"/>
    <param name="laser_z_short"             value="0.1"/>
    <param name="laser_z_max"               value="0.05"/>
    <param name="laser_z_rand"              value="0.05"/>
    <param name="laser_sigma_hit"           value="0.2"/>
    <param name="laser_lambda_short"        value="0.1"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="laser_model_type"          value="likelihood_field"/>

    <param name="odom_model_type"           value="diff"/>
    <param name="odom_alpha1"               value="8.0"/><!-- 回転時のオドメトリの精度が悪いので大きめに設定 -->
    <param name="odom_alpha2"               value="0.2"/>
    <param name="odom_alpha3"               value="0.2"/>
    <param name="odom_alpha4"               value="0.2"/>
    <param name="odom_frame_id"             value="odom"/>
    <param name="base_frame_id"             value="base_link"/>
  </node>

  <!-- NAVIGATION -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
    <rosparam file="$(find ros_navigation)/param_stretch/common_costmap_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find ros_navigation)/param_stretch/common_costmap_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find ros_navigation)/param_stretch/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find ros_navigation)/param_stretch/global_costmap_params_withmap.yaml" command="load" />
    <rosparam file="$(find ros_navigation)/param_stretch/dwa_local_planner_params_fast.yaml" command="load" />
    <rosparam file="$(find ros_navigation)/param_stretch/move_base_params.yaml" command="load" />
    <!-- remap from="/cmd_vel" to="/stretch/cmd_vel" / -->
    <remap from="map" to="map_pp"/>
  </node>

  <!-- 速度スケーリング -->
  <node pkg="ros_navigation" type="speed_scale_controller.py" name="speed_scale_controller" output="screen">
    <remap from="/scaled_cmd_vel" to="/stretch/cmd_vel" />
    <param name="trans_scaling_factor" value="1.3"/>
    <param name="rot_scaling_factor" value="1.4"/>
  </node>

  <!-- VISUALIZE -->
  <node name="rviz" pkg="rviz" type="rviz" output="log" args="-d $(find ros_navigation)/rviz/stretch_navigation.rviz" if="$(arg rviz)" />

</launch>
